Objective-C/CBLCollection.mm

```diff
@@ -21,45 +21,139 @@
 #import "CBLIndexable.h"
 #import "CBLChangeListenerToken.h"
 #import "CBLCollectionChangeObservable.h"
+#import "CBLDatabase+Internal.h"
+#import "CBLStringBytes.h"
+#import "CBLScope.h"
+#import "CBLIndexConfiguration+Internal.h"
+#import "CBLStatus.h"
+#import "CBLScope+Internal.h"
+
+using namespace fleece;
 
 NSString* const kCBLDefaultCollectionName = @"_default";
 
-@implementation CBLCollection
+@implementation CBLCollection {
+    __weak CBLDatabase* _db;
+    C4Collection* _c4col;
+}
 
 @synthesize count=_count, name=_name, scope=_scope;
 
-- (instancetype) initWithName: (NSString*)name
-                        scope: (nullable CBLScope*)scope
-                        error: (NSError**)error {
-    CBLAssertNotNil(name);
+- (instancetype) initWithDB: (CBLDatabase*)db
+             collectionName: (NSString*)collectionName
+                  scopeName: (NSString*)scopeName
+                      error: (NSError**)error {
+    CBLAssertNotNil(db);
+    CBLAssertNotNil(collectionName);
     
     self = [super init];
     if (self) {
-        _name = name;
-        if (scope)
-            _scope = scope;
+        _db = db;
+        _name = collectionName;
+        
+        scopeName = scopeName.length > 0 ? scopeName : kCBLDefaultScopeName;
+        CBLStringBytes cName(collectionName);
+        CBLStringBytes sName(scopeName);
+        C4CollectionSpec spec = { .name = cName, .scope = sName };
+        
+        __block C4Collection* c;
+        __block C4Error err = {};
+        [db safeBlock:^{
+            c = c4db_getCollection(db.c4db, spec);
+            if (!c) {
+                // collection doesn't exists, create one
+                c = c4db_createCollection(db.c4db, spec, &err);
```

Comment:
1. This logic is not correct for the case that you want to get the collection but don't want to create.
2. For the case of create, calling c4db_createCollection() is enough.

I would suggest that changing the signature to something as below, and let the CBLDatabase gives the c4collection.

`(instancetype) initWithDB: (CBLDatabase*)db collection: (C4Collection*)c4col error: (NSError**)error`.

Reply:
CBLDatabase is not the only class which uses the [CBLCollection alloc] code, but CBLScope as well. So I guess, passing the collectionName and ScopeName can make it common NSString -> Slice conversion and everything inside CBLCollection itself.